lib/crypto/src/eddsa.rs

```diff
@@ -0,0 +1,375 @@
+/// fall out of scope.
+#[derive(Copy, Clone, PartialEq)]
+pub(crate) struct ExpandedSecretKey {
+    /// The secret scalar used for signing.
+    pub(crate) scalar: Scalar,
+    /// The domain separator used when hashing the message to generate the
+    /// pseudorandom `r` value.
+    pub(crate) hash_prefix: [u8; 32],
+}
+
+impl ZeroizeOnDrop for ExpandedSecretKey {}
+
+impl ExpandedSecretKey {
+    /// Construct secret key [`Self`] from a byte string of any length.
+    ///
+    /// Secret key will be derived from hashed `bytes`.
+    pub(crate) fn from_bytes(bytes: &[u8]) -> ExpandedSecretKey {
+        let hash = Sha512::default().chain_update(bytes).finalize();
+        let bytes = &*hash;
+        let mut scalar_bytes = [0u8; 32];
+        let mut hash_prefix = [0u8; 32];
+        scalar_bytes.copy_from_slice(&bytes[00..32]);
+        hash_prefix.copy_from_slice(&bytes[32..64]);
+
+        let scalar = Scalar::from_bigint(U256::from_bytes_le(&clamp_integer(
+            scalar_bytes,
+        )));
+        Self { scalar, hash_prefix }
+    }
+}
+
+/// Clamps the given little-endian representation of a 32-byte integer.
+///
+/// Clamping the value puts it in the range:
+/// **n ∈ 2^254 + 8\*{0, 1, 2, 3, ..., 2^251 − 1}**
+///
+/// # Explanation of clamping
+///
+/// For Curve25519, h = 8, and multiplying by 8 is the same as a binary
+/// left-shift by 3 bits. If you take a secret scalar value between 2^251 and
+/// 2^252 – 1 and left-shift by 3 bits, then you end up with a 255-bit number
+/// with the most significant bit set to 1 and the least-significant three bits
+/// set to 0.
+///
+/// The Curve25519 clamping operation takes **an arbitrary 256-bit random
+/// value** and clears the most-significant bit (making it a 255-bit number),
+/// sets the next bit, and then clears the 3 least-significant bits. In other
+/// words, it directly creates a scalar value that is in the right form and
+/// pre-multiplied by the cofactor.
+///
+/// See [clamping reference] for more details.
+///
+/// [clamping reference]: https://neilmadden.blog/2020/05/28/whats-the-curve25519-clamping-all-about/
+#[must_use]
+pub const fn clamp_integer(mut bytes: [u8; 32]) -> [u8; 32] {
+    bytes[0] &= 0b1111_1000;
+    bytes[31] &= 0b0111_1111;
+    bytes[31] |= 0b0100_0000;
+    bytes
+}
+
+impl From<&SecretKey> for ExpandedSecretKey {
+    #[allow(clippy::unwrap_used)]
+    fn from(secret_key: &SecretKey) -> ExpandedSecretKey {
+        let hash = Sha512::default().chain_update(secret_key).finalize();
+        ExpandedSecretKey::from_bytes(hash.as_ref())
+    }
+}
+
+/// Ed25519 signing key which can be used to produce signatures.
+/// Invariant: `verifying_key` is always the public key of
+/// `secret_key`.
+/// This prevents the signing function [oracle attack].
+///
+/// [oracle attack]: https://github.com/MystenLabs/ed25519-unsafe-libs
+#[derive(Copy, Clone, PartialEq)]
+pub struct SigningKey {
```

Comment:
```suggestion
/// Ed25519 signing key which can be used to produce signatures.
///
/// Invariant: `verifying_key` is always the public key of
/// `secret_key`.
/// This prevents the signing function [oracle attack].
///
/// [oracle attack]: https://github.com/MystenLabs/ed25519-unsafe-libs
#[derive(Copy, Clone, PartialEq)]
pub struct SigningKey {
```
avoids https://rust-lang.github.io/rust-clippy/master/index.html#too_long_first_doc_paragraph